### PR TITLE
Add optional expiry slot to Mock transactions

### LINKED ISSUE #1297: Test that we use the right slot number for transaction validation throughout
The edge case we want to test: a transaction is valid when added to the mempool, but has become invalid by the time that the node gets around to producing the block.

(If the mempool would use the wrong slot number to validate it, it wouldn't notice this, and actually produce an invalid block.)

The easiest way to test this is to add a TTL to a transaction.

Note that this is *not* a Mempool test, but a protocol test, as it requires  the Mempool, block forging, and a ticking blockchain time.

This would exercise the chain ticking stuff introduced in #1295.

### PR DESCRIPTION
**Pull request by nfrisby**
Fixes #1297.

Follow-up: `prop_general` is accumulating several block/protocol specific arguments; they probably deserve their own record type/class soon.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs [hunk 1]
```diff
@@ -50,22 +52,31 @@ import           Ouroboros.Consensus.Ledger.Mock.Address
   Basic definitions
 -------------------------------------------------------------------------------}
 
-data Tx = UnsafeTx (Set TxIn) [TxOut]
+data Expiry
+  = DoNotExpire
+  | ExpireAtOnsetOf !SlotNo
+  deriving stock    (Show, Eq, Ord, Generic)
+  deriving anyclass (Serialise, NFData, NoUnexpectedThunks)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs]: Why is `NFData` (and that orphan) needed?

File: ouroboros-consensus/test-consensus/Test/ThreadNet/General.hs [hunk 2]
```diff
@@ -594,3 +598,18 @@ prop_general countTxs k TestConfig{numSlots, nodeJoinPlan, nodeRestarts, nodeTop
         average :: [Double] -> Double
         average [] = 0
         average xs = sum xs / fromIntegral (length xs)
+
+    -- The 'prop_valid_block' argument could, for example, check for no expired
+    -- transactions.
+    --
+    -- TODO apply to all 'TraceForgeAdopted' blocks, not just those in the
+    -- final chain?
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/ThreadNet/General.hs]: Is that possible? Maybe by (ab?)using the tracer you have for `TraceForgeAdopted`?

Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/ThreadNet/General.hs]: Yeah, I was thinking via tracer. It doesn't seem like an abuse, I don't think: the tracer could just accumulate the adopted blocks in a set and check them at the end.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs]: `Tx` derives it, and `Expiry` occurs in `Tx`. I didn't question why `Tx` derives it.

File: ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/TxGen/Mock.hs [hunk 3]
```diff
@@ -40,8 +42,12 @@ genSimpleTx addrs u = do
         outs         = if amount == fortune
             then [outRecipient]
             else [outRecipient, (sender, fortune - amount)]
-    return $ Tx ins outs
+    expiry <- (mkExpiry . SlotNo . fromIntegral) <$> generateBetween 2 10
```
Comment (nfrisby) [inline ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/TxGen/Mock.hs]: This expiry doesn't seem to affect anything else, so 2-10 seems reasonable.

Should we also include no expiry here? If I recall correctly, I only introduced the `DoNotExpire` ctor because I wasn't sure how to update the mempool tests, so I side-stepped it that way: expiry in `ThreadNet`, no expiry in those other tests.

@mrBliss Please advise.

Comment (mrBliss) [reply ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/TxGen/Mock.hs]: I don't think it matters that much. I think I would include `DoNotExpire` here too, just to cover everything.

Should the expiry range be dependent on `k`? Although they are unrelated, it might make a difference when we're testing with network partitions and such? I suppose we can keep it [2, 10] now and update it when/if needed.

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs [hunk 4]
```diff
@@ -330,8 +331,22 @@ instance MockProtocolSpecific c ext => ApplyTx (SimpleBlock c ext) where
 
   type ApplyTxErr (SimpleBlock c ext) = MockError (SimpleBlock c ext)
 
-  applyTx   = const updateSimpleUTxO
-  reapplyTx = const updateSimpleUTxO
+  applyTx           = reapplyTx
+  reapplyTx _ tx tl =
+      case hasExpired of
```
Comment (mrBliss) [inline ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs]: I see CI is failing with a test related to tx/block validation. I believe you also have to take transaction expiration into account when validating blocks (`applyLedgerBlock`). You can move the expiration logic to some common place.

Comment (edsko) [reply ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs]: It shouldn't live in `applyTx` _or_ `applyLedgerBlock`, but rather in the ledger rules.

Comment (mrBliss) [reply ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs]: See Lemma 3 in #1765 :slightly_smiling_face:

**Comment by nfrisby** [comment]
The latest CI failure was due to the unrelated #1779.

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs (hunk 4 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs]: I pushed it into the ledger rules. I introduced the `Ticked` type used in a few modules; should it be renamed and/or located somewhere more general? I think of it (as commented) as a generalization of `TickedLedgerState`. @edsko @mrBliss

File: ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/TxGen/Mock.hs (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/TxGen/Mock.hs]: Done, thanks.

File: ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/BFT.hs [hunk 5]
```diff
@@ -4,19 +4,22 @@ module Test.ThreadNet.BFT (
...
+import qualified Data.Map.Strict as Map
+
 import           Test.QuickCheck
 import           Test.Tasty
 import           Test.Tasty.QuickCheck
 
 import           Cardano.Slotting.Slot
 
+import           Ouroboros.Network.Block (SlotNo (..))
+
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/BFT.hs]: Fixed, thanks.

**Comment by nfrisby** [comment]
I think this is ready for merge, unless you'd like to see changes to the following.

  * the way I pushed the logic further into the ledger rule code
  * the `Ticked` type I used to do so

If not, I think this is ready for `bors`.

**Comment by nfrisby** [comment]
@edsko and I had a call. He pointed out a slightly higher-level location in which to check for expiry. Moving it there let me remove `Ticked`.

In order to relocate the check, I also simplified `HasUtxo` to `HasTxs` -- I need to access the actual `Tx`s in order to check them for expiry. Is this too much of a concretization? I suppose I could add something about expiry to `HasUtxo`, but this simplification seems fine in the mock ledger code.

**Comment by nfrisby** [comment]
I just now rebased onto `origin/master`.

**Comment by nfrisby** [comment]
I just now rebased onto `origin/bors/staging`, to address the conflicts from the new `PropGeneralArgs`.

**Comment by nfrisby** [comment]
I rebased on `origin/master`.

File: ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/BFT.hs [hunk 6]
```diff
@@ -44,11 +46,30 @@ tests = testGroup "BFT"
...
+          , initSeed     = Seed (6358650144370660550,17563794202468751585,17692838336641672274,12649320068211251815,18441126729279419067)
+          }
+    , testProperty "simple convergence" $ \tc ->
+        -- TODO k > 1 as a workaround for Issue #1511.
+        --
         forAll (SecurityParam <$> elements [2 .. 10]) $ \k ->
         prop_simple_bft_convergence k tc
     ]
+  where
+    defaultSlotLengths = singletonSlotLengths $ slotLengthFromSec 1
```
Comment (mrBliss) [inline ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/BFT.hs]: Does the repro depend on the exact slot length?

I'm inclined to use 20s here, because that's the default slot length for BFT.

Also, you could use `defaultSlotLengths` in the first test case too.

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/UTxO.hs [hunk 7]
```diff
@@ -89,61 +104,63 @@ data UtxoError
...
-  txOuts     = foldr (Map.union . txOuts)    Map.empty
-  confirmed  = foldr (Set.union . confirmed) Set.empty
-  updateUtxo = repeatedlyM updateUtxo
-
-instance HasUtxo a => HasUtxo (Chain a) where
-  txIns      = txIns      . toOldestFirst
-  txOuts     = txOuts     . toOldestFirst
-  updateUtxo = updateUtxo . toOldestFirst
-  confirmed  = confirmed  . toOldestFirst
+class HasTxs a where
```
Comment (mrBliss) [inline ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/UTxO.hs]: There already exists such a class: https://github.com/input-output-hk/ouroboros-network/blob/cca25adb2cc393177286feaba4c503a033343c36/ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs#L93

Either we reuse that class (maybe generalise it to `a` instead of `blk`), and if that's not possible (`Tx` vs `GenTx`), rename this new class to avoid confusion.

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/State.hs [hunk 8]
```diff
@@ -75,13 +82,32 @@ updateMockTip hdr (MockState u c t)
...
+                -> Tx
+                -> MockState blk
+                -> Except (MockError blk) (MockState blk)
+updateMockUTxO1 now tx (MockState u c t) = case hasExpired of
+    Just e  -> throwError e
+    Nothing -> do
+      u' <- withExcept MockUtxoError $ updateUtxo tx u
+      return $ MockState u' (c `Set.union` confirmed tx) t
+  where
+      Tx expiry _ins _outs       = tx
```
Comment (mrBliss) [inline ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/State.hs]:
```suggestion
      Tx expiry _ins _outs = tx
```

Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/State.hs]: Done, thanks.

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/UTxO.hs (hunk 7 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/UTxO.hs]: I renamed the new one; the two have slightly different semantics (going by the existing class's comment). Moreover, this new one is specific to the mock Tx -- I don't think generalizing the existing one would be trivial and/or as clear.

Done, thanks.

File: ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/BFT.hs (hunk 6 above)
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/BFT.hs]: I switched it to 20s. Done, thanks.

**Comment by nfrisby** [comment]
Update: I addressed the comments and rebased onto master.
